### PR TITLE
Add factory extension and traits API

### LINKED ISSUE #18: Composable Factories?
Is it possible to do composition for Factories in here?

Since FactoryBot is a point of reference for this project, I'm looking for something like `traits`, where I can generate variations from a "Base Factory".

### PR DESCRIPTION
**Pull request by stevehanson**
A limitation brought up in https://github.com/thoughtbot/fishery/issues/18 is that there was not an obvious way to extend factories or to simplify building complex objects.

Up till now, if we wanted to create a convenient way to build an object, it was common to resort to transient parameters and use them like this:

```typescript
userFactory.build({ name: 'Katja' }, { transient: { admin: true, withPosts: 2 })
```

The above `build` feels clunky and quickly gets unwieldy when creating complex objects. The factory definition is also complicated by the use of the transient params. 

This PR introduces an API for extending factories by building up defaults for its params, transient params, associations, and callbacks, so we could, for example, extend the `userfactory` to create an `adminFactory`. This new API also allows for creating convenience builder methods on existing factories (like factory_bot's "traits") which allows us to create and use APIs like:

```typescript
userFactory.admin().withPosts(2).build()
```

### Extending factories

This adds four factory "extension methods": `params`, `transient`, `associations`, and `afterCreate`. These set default attributes that get passed to the factory on `build`:

```typescript
const userFactory = Factory.define<User>(() => ({
  name: 'Kassandra',
  admin: false,
}));

const adminFactory = userFactory.params({ admin: true });
const admin = adminFactory.build();
admin.admin; // true
```

The extension methods return a new factory and don't mutate the factory they are called on. Extension methods can also be chained:

```typescript
const eliFactory = userFactory
  .params({ admin: true })
  .params({ name: 'Eli' })
  .afterCreate(user => console.log('hello'))
  .afterCreate(user => console.log('there'));

const user = eliFactory.build();
// log: hello
// log: there
user.name; // Eli
user.admin; // true
```

The params passed to `build` still take precedence over those defined when extending the factory:

```typescript
const user2 = eliFactory.build({ admin: false });
user.name; // Eli
user2.admin; // false
```

### Builder methods (traits)

The factory extension API paved the way for adding what I call "builder methods" to factories. Since factories are just classes, you can now subclass `Factory` and define any methods on a factory. These methods can use the extension methods to create builders:

```typescript
class UserFactory extends Factory<User, Factories, UserTransientParams> {
  admin(adminId = '') {
    return this.params({
      admin: true,
      adminId: adminId || `admin-${this.sequence()}`,
    });
  }

  registered() {
    return this
      .params({ memberId: this.sequence() })
      .transient({ registered: true })
      .associations({ profile: factories.profile.build() })
      .afterCreate(user => console.log(user))
  }
}

// instead of Factory.define<User>
const userFactory = UserFactory.define(() => ({ ... }))

const user = userFactory.admin().registered().build()
```

Prior to this PR, there was not a way to subclass `Factory` because the static `Factory.define` method was not typed in a way that it would work polymorphically (it always returned a `Factory`). This adds power to customize factories however users prefer.

Closes #18.

### CODE REVIEW

**Comment by ObliviousHarmony** [comment]
I wonder if now might be a good time to change `afterCreate` to `afterBuild`? If you're expanding the API with the `afterCreate` method, it might be better to settle on a naming that won't conflict with async `create` functionality?

**Comment by stevehanson** [comment]
[Traits / extending factories](https://trello.com/c/xryQVmyF/21-traits-extending-factories)

**Comment by uchagani** [comment]
Can this be merged?

File: README.md [hunk 1]
```diff
@@ -290,6 +294,90 @@ export default Factory.define<User, Factories>(
...
+If you find yourself frequently building objects with a certain set of
+properties, it might be time to either extend the factory or create a
+reusable builder method.
+
+Factories are just classes, so adding reusable builder methods is as simple
+as subclassing `Factory` and defining any desired methods:
+
+```typescript
+class UserFactory extends Factory<User, Factories, UserTransientParams> {
+  admin(adminId = '') {
```
[the remark this answers is not in the document]
Comment (stevehanson) [reply README.md]: Good catch. Updated in e28ea4d.

**Pull request merged**